### ISSUE TITLE
[question] How to list all OpenGL ES 2 functions?

### ISSUE THREAD

**Issue by yurivict** [KhronosGroup/OpenGL-Registry#279]
The XML file https://github.com/KhronosGroup/OpenGL-Registry/blob/master/xml/gl.xml contains the list of all OpenGL functions.

Is there a way to find a sublist of OpenGL ES 2 functions without explicitly listing them?

**Comment by oddhack** [comment]
Sure. Look at the \<feature> block corresponding to ES 2, the one with name="GL_ES_VERSION_2_0" and number="2.0". That doesn't include ES2-specific extensions, which are potentially the \<extension> blocks with supported= attribute containing "gles2" (although some of those extensions may only work against ES 3 or later).

**Comment by oddhack** [comment]
N.b. the header generation is driven off of gl.xml via a frontend script, genheaders.py, which selects interfaces from the XML depending on the header being generated. It's a little indirect but still straightforward to reverse-engineer which interfaces belong with which API.

**Comment by yurivict** [comment]
Thank you for this information!

I have one more question:
For my project I need to have every function that is used in WebGL as supported by browsers. I read somewhere that ```WebGL``` corresponds to ```OpenGL ES 2```. My question is what exactly is the ```WebGL``` set?

For example, the extension ```GL_NV_non_square_matrices``` supports ```gles2``` and has functions like ```glUniformMatrix4x2fvNV``` which look is NVidia-specific. Are such NVidia-specific extension functions part of WebGL?

**Comment by oddhack** [comment]
You'll be best off asking this on one of the WebGL-specific forums or mailing lists (see https://www.khronos.org/webgl/wiki/Main_Page). WebGL is based on ES 2, but it is a different API and has a Javascript binding, often with very different parameters (for example, passing in image objects instead of pointers to image data for the texture commands). We don't describe the WebGL API in gl.xml.

There are not many WebGL extensions, compared to OpenGL - see their registry at https://www.khronos.org/registry/webgl/extensions/ . I believe that's primarily because WebGL is aimed at enabling maximum portability of apps across browsers and implementations. In particular there are virtually no vendor-specific WebGL extensions, AFAIK. Regarding NPOT textures in particular, see the first item in https://www.khronos.org/webgl/wiki/WebGL_and_OpenGL_Differences .

**Comment by yurivict** [comment]
I asked my question on the WebGL forum: https://community.khronos.org/t/is-there-an-xml-definition-of-webgl-api-and-its-relationship-to-opengl-es2/104177

Thank you for your help!

**Comment by yurivict** [comment]
I think I figured this issue out myself, as much as I could.

OpenGL specs are very nice, they are in a single ```gl.xml``` file.

However, WebGL isn't defined this way. WebGL specs are defined in two html files: https://github.com/KhronosGroup/WebGL/blob/master/specs/1.0.3/index.html and https://github.com/KhronosGroup/WebGL/blob/master/specs/2.0.0/index.html

The WebGL project parses html files using the python script (!!!) and generates .IDL files. As it turns out, these html files are malformatted, ```xmllint``` complains about that, and it looks like these files are hand-written and their formatting isn't consistent across functions that they describe. This makes reading them quite difficult.

Basically, WebGL specs don't exist in well formed, machine-readable form, sadly. This is an unpleasant surprise.
